### PR TITLE
update tensor indexing to fail on `non_xla_tensor[xla_tensor]`

### PR DESCRIPTION
**Pull request by bdhirsh**
pytorch core PR: https://github.com/pytorch/pytorch/pull/69607#discussion_r821732328

As part of porting `at::Tensor::index()` to be a structured kernel in core, we're making it so that you can't index a cpu tensor with non-cpu indices (this is BC-breaking but from the linked discussion above it was agreed that realistically nobody should be doing this).

This PR adds a corresponding check in the XLA kernel for `index()` to check for the same thing.

### CODE REVIEW

File: torch_xla/csrc/aten_xla_type.cpp [hunk 1]
```diff
@@ -1593,6 +1593,15 @@ at::Tensor XLANativeFunctions::index(
     const at::Tensor& self,
     const c10::List<c10::optional<at::Tensor>>& indices) {
   XLA_FN_COUNTER("xla::");
+  bool indices_on_cpu_or_dev = std::all_of(
+      indices.begin(), indices.end(), [=](const at::OptionalTensorRef& opt) {
+        return opt.has_value() ? (opt->is_cpu() || opt->device() == dev) : true;
```
Comment (JackCaoG) [inline torch_xla/csrc/aten_xla_type.cpp:1598]: seems like you miss the `dev` init>

**Comment by bdhirsh** [comment]
Hey @JackCaoG, I've heard that there's a new workflow for merging XLA + core dependent patches where we basically merge the XLA-side PR first (and then updated a .github/xla_commit_hash.txt` file in the core-side PR before merging that. cc @janeyx99)

This PR should be ready, since the XLA tests are passing on the pytorch core PR. Can I get an approve?

**Comment by JackCaoG** [comment]
I saw 
```
ERROR: test_advancedindex_mixed_cpu_devices_xla (__main__.TestDevicePrecisionXLA)
```
was failing. Is it because you remove the torch pin?

**Review by JackCaoG — APPROVED** [review]
Thanks

**Pull request merged**